## Part 1: Preprocessing

In [22]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [23]:
# Determine the number of unique values in each column
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [24]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[['Attrition', 'Department']]

In [25]:
# Create a list of at least 10 column names to use as X data
columns = ['Age', 'BusinessTravel', 'Education', 'JobLevel', 'Education', 
           'YearsAtCompany','YearsWithCurrManager', 'YearsSinceLastPromotion', 'DistanceFromHome', 'JobInvolvement']

# Create X_df using your selected columns
X_df = attrition_df[columns]

# Show the data types for X_df
X_df.dtypes


Age                         int64
BusinessTravel             object
Education                   int64
JobLevel                    int64
Education                   int64
YearsAtCompany              int64
YearsWithCurrManager        int64
YearsSinceLastPromotion     int64
DistanceFromHome            int64
JobInvolvement              int64
dtype: object

In [26]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, random_state=42)

In [27]:
# Convert your X data to numeric data types however you see fit
X_train = pd.get_dummies(X_train)
X_test = pd.get_dummies(X_test)


In [28]:
# Create a StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler to the training data
x_train_scaled = scaler.fit_transform(X_train)

# Scale the training and testing data
x_train_scaled = scaler.transform(X_train)
x_test_scaled = scaler.transform(X_test)

In [29]:
from sklearn.preprocessing import OneHotEncoder

# Create a OneHotEncoder for the Department column
encoder = OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
encoder.fit(y_train[['Department']])

# Create two new variables by applying the encoder
# to the training and testing data
var1 = encoder.transform(y_train[['Department']])
var2 = encoder.transform(y_test[['Department']])
print(var1)
print(var2)

[[0. 1. 0.]
 [0. 0. 1.]
 [0. 0. 1.]
 ...
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
[[0. 0. 1.]
 [0. 1. 0.]
 [1. 0. 0.]
 ...
 [0. 1. 0.]
 [1. 0. 0.]
 [0. 1. 0.]]


In [30]:
# Create a OneHotEncoder for the Attrition column
encoder_2 = OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
encoder_2.fit(y_train[['Attrition']])

# Create two new variables by applying the encoder
# to the training and testing data


var3 = encoder_2.transform(y_train[['Attrition']])
var4 = encoder_2.transform(y_test[['Attrition']])
print(var3)
print(var4)

[[1. 0.]
 [1. 0.]
 [1. 0.]
 ...
 [0. 1.]
 [1. 0.]
 [1. 0.]]
[[1. 0.]
 [1. 0.]
 [0. 1.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [0. 1.]
 [0. 1.]
 [1. 0.]
 [0. 1.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [0. 1.]
 [1. 0.]
 [1. 0.]
 [0. 1.]
 [0. 1.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [0. 1.]
 [1. 0.]
 [0. 1.]
 [0. 1.]
 [0. 1.]
 [0. 1.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [0. 1.]
 [0. 1.]
 [1. 0.]
 [0. 1.]
 [0. 1.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [0. 1.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [1. 0.]
 [0. 1.]
 [1. 0.]
 [1.

## Part 2: Create, Compile, and Train the Model

In [31]:
# Find the number of columns in the X training data.
input_shape = X_train.shape[1]

# Create the input layer
input_layer = layers.Input(shape=(input_shape,))

# Create at least two shared layers
shared_layer_1 = layers.Dense(128, activation='relu')
shared_layer_2 = layers.Dense(64, activation='relu')

In [32]:
# Create a branch for Department
# with a hidden layer and an output layer
department = shared_layer_1(input_layer)
# Create the hidden layer
department = shared_layer_2(department)

# Create the output layer
department = layers.Dense(3, activation='softmax', name='Department')(department)

In [33]:
# Create a branch for Attrition
# with a hidden layer and an output layer
attrition = shared_layer_1(input_layer)
# Create the hidden layer
attrition = shared_layer_2(attrition)

# Create the output layer
attrition = layers.Dense(2, activation='sigmoid', name='Attrition')(attrition)

In [34]:
# Create the model
model = Model(inputs=input_layer, outputs=[department, attrition]) 

# Compile the model
model.compile(
    optimizer='adam',
    loss={
        'Department': 'categorical_crossentropy',
        'Attrition': 'categorical_crossentropy'
    },
    metrics={
        'Department': ['accuracy'],  
        'Attrition': ['accuracy']    
    }
)
# Summarize the model
model.summary()

Model: "functional_1"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_1       │ (None, 12)        │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_2 (Dense)     │ (None, 128)       │      1,664 │ input_layer_1[0]… │
│                     │                   │            │ input_layer_1[0]… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_3 (Dense)     │ (None, 64)        │      8,256 │ dense_2[0][0],    │
│                     │                   │            │ dense_2[1][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Department (Dense)  │ (None, 3)         │        195 │ dense_3[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Attrition (Dense)   │ (None, 2)         │        130 │ dense_3[1][0]     │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 10,245 (40.02 KB)

 Trainable params: 10,245 (40.02 KB)

 Non-trainable params: 0 (0.00 B)

In [35]:
# Convert y_train and y_test to numeric data types using one-hot encoding
y_train_department = encoder.transform(y_train[['Department']])
y_test_department = encoder.transform(y_test[['Department']])
y_train_attrition = encoder_2.transform(y_train[['Attrition']])
y_test_attrition = encoder_2.transform(y_test[['Attrition']])

# Train the model
model.fit(
    x_train_scaled,
    {'Department': y_train_department, 'Attrition': y_train_attrition},
    epochs=10,
    batch_size=32,
    shuffle=True,
    verbose=1
)


Epoch 1/10
35/35 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - Attrition_accuracy: 0.7752 - Attrition_loss: 0.5297 - Department_accuracy: 0.5313 - Department_loss: 0.9684 - loss: 1.4983   
Epoch 2/10
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - Attrition_accuracy: 0.8196 - Attrition_loss: 0.4480 - Department_accuracy: 0.6317 - Department_loss: 0.7907 - loss: 1.2384 
Epoch 3/10
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - Attrition_accuracy: 0.8533 - Attrition_loss: 0.3894 - Department_accuracy: 0.6490 - Department_loss: 0.7574 - loss: 1.1469 
Epoch 4/10
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - Attrition_accuracy: 0.8355 - Attrition_loss: 0.4148 - Department_accuracy: 0.6759 - Department_loss: 0.7094 - loss: 1.1241 
Epoch 5/10
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - Attrition_accuracy: 0.8346 - Attrition_loss: 0.4005 - Department_accuracy: 0.6537 - Department_loss: 0.7270 - loss: 1.1273 
Epoch 6/10
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - Attrition_accuracy: 0.8478 - Attrition_loss: 0.3909 - Department_a

In [36]:
# Evaluate the model with the testing data
model_loss, department_loss, attrition_loss, department_accuracy, attrition_accuracy = model.evaluate(
    x_test_scaled,
    {'Department': y_test_department, 'Attrition': y_test_attrition},
    verbose=2
)

12/12 - 0s - 14ms/step - Attrition_accuracy: 0.8696 - Attrition_loss: 0.3544 - Department_accuracy: 0.6495 - Department_loss: 0.7940 - loss: 1.1286


In [37]:
# Print the accuracy for both department and attrition
print(f"Department Accuracy: {department_accuracy}")
print(f"Attrition Accuracy: {attrition_accuracy}")

Department Accuracy: 0.8695651888847351
Attrition Accuracy: 0.6494565010070801


In [38]:
# Check the class distribution for Attrition
print("Attrition class distribution:")
print(y_df['Attrition'].value_counts())

# Check the class distribution for Department
print("\nDepartment class distribution:")
print(y_df['Department'].value_counts())

Attrition class distribution:
Attrition
No     1233
Yes     237
Name: count, dtype: int64

Department class distribution:
Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. In this case, the imbalince in the data between no/yes of the attribute class may lead to misrepresentation of the quality of the model if only evaluating accuracy.  It may be a good starting place but I would consider using precision as well.
2. I used softmax and sigmoid, chosen for the multiclass and binary classification.  I started with softmax for both and it works but I suspect using a binary classification function makes more sense on a binary output.
3. Considering the imbalance in the data, more data would have helped. I also already changed to a sigmoid activation function to simplify the model for use on binary classification.  While I don't see it in this case, a dropout function could be used to overfit.  Again, on this case with a .68 and loss of over 1, I don't necessarily see a need.  Looking at the F1-score could help due to the imbalanced data.